### ISSUE TITLE
Handle zsh syntax

### ISSUE THREAD

**Issue by gibfahn** [bash-lsp/bash-language-server#252]
I have a bunch of scripts (not least my dotfiles) where I have to deal with both bash and zsh. However whenever I add some zsh syntax, for example `[[ -n "${foo[(r)b]}" ]] && echo 'b was found.' || echo 'b was not found.'`, the entire file becomes highlighted as an error.

I guess possible solutions are:

1. Support zsh syntax
2. Limit the error highlighting to just the affected line
3. Recognise a magic comment to ignore issues for a particular line.

The first option sounds like a big rewrite, to the point where it's probably a different project (although supporting both would be really useful as scripts often mingle the two formats).

Unsure if the second and third options are feasible.

**Comment by zfogg** [comment]
+1

**Comment by pboushy** [comment]
Would love to see this too.

**Comment by lukelbd** [comment]
Just wanted to point out a similar thread at koalaman/shellcheck#809. Personally these two issues are the only thing keeping me from switching to zsh (I use vim with [ale](https://github.com/dense-analysis/ale) and [vim-lsp](https://github.com/prabirshrestha/vim-lsp)).

**Comment by skovhus** [comment]
Just a little update here. Unfortunately we are still blocked by having a better parser for zsh. And official shellcheck support would also be great. 

But the experience for zsh files should have been improved compared to when this issue originally was created. The language server doesn’t highlight any parsing issues anymore.

**Comment by gibfahn** [comment]
I saw https://github.com/koalaman/shellcheck/issues/809#issuecomment-1411855521 , which points to https://github.com/psprint/zsh-sweep . I haven't tried it, but if it works well for zsh files, integrating that instead of shellcheck for zsh might be the easiest path forward.

**Comment by hinell** [comment]
I don't think this is the right place for a such request. I think we should create a completely separate LSP for the zsh. If anyone is interested in porting zsh parser & lexer to the new LSP please let me know. Let's unite to create it.

I would expected it to be written by using C/C++/Clang/Cmake/ and other building tools.

**Comment by hinell** [comment]
We need to port ZSH parser & lexer  to a new LSP server.

If anyone is interested in porting zsh parser & lexer to the new LSP please let me know. Let's unite to create it.

I would expected it to be written by using C++/Clang/Cmake/ and other building tools. Currently ZSH uses GNU Autoconfig and some other old and outdated stuff.

**Comment by hinell** [comment]
@Freed-Wu This is a huge mistake. These flags shouldn't be a part of bash syntax, cause it's likely to fail in real world dev tools.

FYI @amaanq is currently busy with writing zsh parser from scratch. Get in touch with him if anyone interested. 

I think the best approach would be adopting already-existing zsh parser and provide tree-sitter api instead. It would be much better.

**Comment by kolchurinvv** [comment]
so seeing how this issue is about to celebrate it's 4th birthday, safe to say zsh support isn't coming?

**Comment by chrisgrieser** [comment]
@kolchurinvv the LSP can at least partially be used for zsh, due to its similarity to bash. I've been using this LSP mostly for zsh files, and stuff like variable renaming or symbols do work. You only have to force it to work with zsh files in your lsp configuration. (I use nvim where I just force zsh files to be interpreted as zsh for that.)

Diagnostics via shellcheck do not work though, but there is an open PR that allows you to configure the LSP to have shellcheck work on zsh files as well: https://github.com/bash-lsp/bash-language-server/pull/1133
There is also an open PR

**Comment by hinell** [comment]
@kolchurinvv ZSH is drastically different from Bash. We would need a separate ZSH lsp. Too many subtle points. Bash and zsh syntaxes aren't mutually compatible and even behave differently.

**Comment by kolchurinvv** [comment]
Fair enough. Why not close this issue then with "won't implement" or something? 🙂

**Comment by 0xdevalias** [comment]
> FYI @amaanq is currently busy with writing zsh parser from scratch. Get in touch with him if anyone interested.

I was curious where this zsh parser work was happening. From @amaanq's profile:

> I maintain and contribute to various tree-sitter repositories, including all the [upstream](https://github.com/tree-sitter) ones, and those under the [tree-sitter-grammars](https://github.com/tree-sitter-grammars) org.

Within https://github.com/tree-sitter-grammars:

- https://github.com/tree-sitter-grammars/tree-sitter-zsh
  - > Zsh grammar for `tree-sitter`

I'm not sure if this was the parser work that was being referred to, but figured I would link it here in case it was useful for this issue / related knowledge.

**Comment by hinell** [comment]
@0xdevalias Thanks for reaching us out once again. It seems that the development on zsh-grammar for `tree-sitter` has staled for sure, but I think it would be best basis for future `zsh-lsp`. I'm still convinced that providing support for ZSH in `bash-language-server` would be too expensive and error-prune.

**Comment by skovhus** [comment]
> I'm still convinced that providing support for ZSH in bash-language-server would be too expensive and error-prune.

You might be right. It would be interesting to spike how manageable it would be to make the tree sitter parser (bash or zsh) pluggable in the bash language server. The number of places we need to branch out could determine if it would make sense to create a new LSP server or integrate zsh with bash language server.

**Comment by chrisgrieser** [comment]
Even without a treesitter parser for zsh, I am already using the bash language server with zsh just fine. Since zsh and bash use a very similar syntax, it works in 90% of my zsh files without any issues. Only thing missing are the diagnostics, which is pending the PR I made in #1133.

**Comment by hinell** [comment]
@chrisgrieser 
>_very similar_

But not the same. You use `bash-lsp` as long as you write scripts compatible with both shells. As I said earlier: otherwise there are too many subtle differences in both syntax and behavior. 

>_diagnostics_

If diagnostics were implemented by `bash-language-server` the latter would complain about every line lacking `;` at the end in zsh scripts which is usually permissible in zsh, therefore, forcing you to actually write `bash` and not `zsh`.  That's not what I would expect.
